Build SDK Gutenpack: Use file-loader to include images

sirreal opened the pull request:
#### Changes proposed in this Pull Request

* Add `.png`, `.jpg` and `.gif` files to the file loader config
* Add hash to file loader name
* Use the file loader to bundle Simple Payments block static assets
* Updates the jetpack-blocks scripts to
  - Add clean (removes build
  - Clean on publish

- [x] Requires a Jetpack change to include these assets: https://github.com/Automattic/jetpack/pull/10570

#### Testing instructions

- Ensure related posts placeholders continue to work well in WP Admin (Add Gutenberg related posts block)
- Ensure related posts placeholders continue to work well in Gutenberg wpcalypso (Add Gutenberg related posts block)
- Ensure related posts demo content continue to work well in Calypso (related posts settings for a Jetpack site) - https://calypso.live/settings/traffic/?branch=update/import-static-assets-gberg-simplepay
* Simple payments button is correctly bundled and loaded in wpcalypso
* Simple payments button is correctly bundled and loaded in Jetpack site WP Admin (https://jurassic.ninja/create?jetpack-beta&branch=update/gutenpack-include-full-build&shortlived&wp-debug-log&gutenberg&gutenpack&calypsobranch=update/import-static-assets-gberg-simplepay
)
* Do a publish dry run and verify produced files looks correct:
  ```sh
  cd client/gutenberg/extensions/presets/jetpack
  npm publish --dry-run
  ```

Something like this:

```
npm notice === Tarball Contents ===
npm notice 912B    package.json
npm notice 508B    README.md
npm notice 129B    build/block-manifest.json
npm notice 7.8kB   build/editor-beta.css
npm notice 178.9kB build/editor-beta.js
npm notice 7.8kB   build/editor-beta.rtl.css
npm notice 6.8kB   build/editor.css
npm notice 171.0kB build/editor.js
npm notice 6.8kB   build/editor.rtl.css
npm notice 7.5kB   build/images/paypal-button.png
npm notice 8.2kB   build/images/paypal-button@2x.png
npm notice === Tarball Details ===
npm notice name:          @automattic/jetpack-blocks
npm notice version:       6.8.0-beta.1
npm notice package size:  136.5 kB
npm notice unpacked size: 396.5 kB
npm notice shasum:        cd6e15374eec0316ebf74262f09e0f396b7e2fda
npm notice integrity:     sha512-x7bx+oSQ+2WTQ[...]RDlYSj80cxqjg==
npm notice total files:   11
npm notice
+ @automattic/jetpack-blocks@6.8.0-beta.1
```

Supersedes https://github.com/Automattic/wp-calypso/pull/28360 and https://github.com/Automattic/jetpack/pull/10559

Fixes #27838 (cc: @jeffersonrabb, please verify)

FYI: @Automattic/luna @Automattic/poseidon @Automattic/calypso-sdk

sirreal committed:
Add png handling to file loader

sirreal committed:
Use file-loader for simple-payments paypal pngs

sirreal committed:
Add clean script to jetpack-blocks, run on publish

simison commented on client/gutenberg/extensions/presets/jetpack/package.json:
```diff
@@ -10,8 +10,9 @@
 		"access": "public"
 	},
 	"scripts": {
+		"clean": "rm -fr ./build",
```
Just noting that this isn't Windows friendly but I'm okay with that since this isn't the main `package.json` for Calypso.

If we use Lerna more extensively, this might come back to us, though.

simison commented on client/gutenberg/extensions/simple-payments/product-placeholder.jsx:
```diff
@@ -9,53 +9,48 @@ import { __ } from 'gutenberg/extensions/presets/jetpack/utils/i18n';
  */
 import './product-placeholder.scss';
+import paypalImage from './paypal-button.png';
+import paypalImage2x from './paypal-button@2x.png';
```
😍

enejb replied:
Could this be something that is handled via the build script? 
Creating 2x images.

simison replied:
Good idea — definitely material for the backlog.

sirreal replied:
We don't want to be creating 2x images from 1x 😉 . Anything automated would preferably scale _down_.

simison commented on webpack.config.js:
```diff
@@ -236,7 +236,7 @@ function getWebpackConfig( { cssFilename, externalizeWordPressPackages = false }
 				},
 				{
-					test: /\.(svg)$/,
+					test: /\.(?:svg|png)$/i,
```
I think we might as well add `jpg` here already now so that the next one who needs it wouldn't need to wonder why it isn't working for them. E.g. Map and latest posts blocks both use jpegs.

sirreal replied:
I was going to start adding a lot but wasn't sure where to draw the line…

`svg|png|gif|jpg|jpeg|ico|png|tiff|webp`

Add em as you need em?

sirreal replied:
> Map and latest posts blocks both use jpegs.

I'll role those changes in here.

simison replied:
Just adding `jpg` seems like a pragmatic and useful choice right now.

sirreal replied:
Done, related posts updated 🎉

sirreal committed:
Add jpegs to file-loader

sirreal committed:
Load related-posts placeholders with file-loader

sirreal commented:
@tyxla - With the hardcoded paths to Calypso assets in the previous (beta bundle) related posts, with this PR those links would break. The assets are not at the same URL. Is that a concern?

tyxla commented on client/my-sites/site-settings/related-posts/related-content-preview.jsx:
```diff
@@ -11,11 +11,14 @@ import { localize } from 'i18n-calypso';
  * Internal dependencies
  */
 import FormLabel from 'components/forms/form-label';
+import imgCatBlog from 'gutenberg/extensions/related-posts/cat-blog.png';
```
The single fact that we have to import images in site settings from our Gutenberg extensions is enough reason to NOT move those images there, but keep them where they were before. Not to mention that we might be using these somewhere else, so I'd refrain from moving them.

tyxla commented:
@sirreal I disagree with moving the related posts images to the Gutenberg extensions directory, because they're used on other places too. Also see https://github.com/Automattic/wp-calypso/pull/28378#issuecomment-436966057

sirreal commented:
> The single fact that we have to import images in site settings from our Gutenberg extensions is enough reason to NOT move those images

@tyxla Sure, I agree. Where should we put them? There's no obvious place to share these outside of the public directory. The public directory isn't ideal with the file loader.

I don't care too much where the images go. I'd like to use the same images in both places like we're currently doing, but that's not necessary. Moving them into `client` seems like a necessity because that's where webpack is configured to resolve imports.

If they don't move into client, we'd have to do relative imports everywhere:

```js
import imgCatBlog from '../../../../public/images/related-posts/cat-blog.png';
```

That doesn't seem like a good approach.

> Not to mention that we might be using these somewhere else, so I'd refrain from moving them.

Should we duplicate them and leave them where they are in `public` indefinitely? I took a quick look at Jetpack and these files don't seem to be used there. I hope other projects are not depending un public Calypso assets. That seems like a bad practice.

simison commented:
>I hope other projects are not depending un public Calypso assets. That seems like a bad practice.

Calypso-Desktop might do just that. :-/

simison commented:
I'd say if we want extra time to think how to do about related-posts assets, we could merge this without those changes and loop back to it another week.

sirreal commented:
> Calypso-Desktop might do just that. :-/

They're bundled with the app and served from URLs in electron like http://127.0.0.1:41050/calypso/images/related-posts/cat-blog.png

simison reviewed (approved):
I built the bundle using this command:
```
npm run sdk -- gutenberg \
  ./client/gutenberg/extensions/presets/jetpack \
  --output-dir=../../jetpack/node_modules/@automattic/jetpack-blocks/build
```
...and ran `yarn build` in Jetpack.

:thumbsup: Blocks load in the editor with Paypal png coming from:
```
http://localhost/wp-content/plugins/jetpack/_inc/blocks/images/paypal-button@2x.png
```

I tested also with `NODE_ENV=production` when building with SDK.

:thumbsup: When building Calypso and placing simple payments block in Gutenberg, the file loads from:
```
http://calypso.localhost:3000/calypso/images/paypal-button@2x.png
```

jeffersonrabb commented on webpack.config.js:
```diff
@@ -236,7 +236,7 @@ function getWebpackConfig( { cssFilename, externalizeWordPressPackages = false }
 				},
 				{
-					test: /\.(svg)$/,
+					test: /\.(?:jpg|jpeg|png|svg)$/i,
```
Why no `.gif`?

sirreal replied:
De we need it? Happy to add it if so 👍

jeffersonrabb replied:
I don't need a .gif at this moment, but I can imagine plenty of uses for animated gifs in blocks. Just seems like something that will mess someone up, someday, so unless there's a downside might as well support it out of the gate?

tyxla commented:
> Moving them into client seems like a necessity because that's where webpack is configured to resolve imports.

Images are a kind of a special case, considering that we have many of them outside of client. Perhaps add a resolve exception for them?

> Should we duplicate them and leave them where they are in public indefinitely?

I'd prefer not having to duplicate anything, ever.

> If they don't move into client, we'd have to do relative imports everywhere:

That's not a good approach indeed, let's try to avoid that.

> I took a quick look at Jetpack and these files don't seem to be used there. I hope other projects are not depending on public Calypso assets.

Truth is, we can't be sure they're not used anywhere else 😞

jeffersonrabb reviewed (dismissed):
There's potential for filename conflict here. Since all the blocks in a preset now share a single `images` folder in the build directory, if multiple blocks have identical assets (e.g. `shim.png` or `icon.svg`) the last one the build script processes wins. The blocks earlier on in the build process will be using an incorrect asset. Maybe in the build directory, each block should have a subdirectory within images, e.g.?

<img width="251" alt="screen shot 2018-11-08 at 8 53 25 am" src="https://user-images.githubusercontent.com/1477002/48202811-e59e6380-e333-11e8-954d-417b65a8ce02.png">

Or each block has a directory at the root of the build directory following the recent change in block-specific view files?

sirreal commented:
> There's potential for filename conflict here.

Good point, although that's nothing new. @blowery mentioned via another channel starting to use the `[hash]` file-loader option. That should provide cache-busting as he mentioned, but is also a good way to protect against conflicts.

We can add that here.

sirreal committed:
Add hash to file-loader

sirreal committed:
Add gif to file-loader

sirreal commented:
Added gif support and hash to name. Filenames now look like:

`/images/paypal-button-1e53882e702881f8dfd958c141e65383.png`

jeffersonrabb commented:
> Added gif support and hash to name

Tested both, works well!

jeffersonrabb reviewed (approved):
Looks good. Tried it out with Map block and everything I needed worked perfectly.

sirreal committed:
Move shared images into client/assets/images

sirreal committed:
Restore images in public folder

sirreal commented:
Rebased.

I've restored the related posts images to their original location so they continue to be available.

I've duplicated the images in a new directory `client/assets/images/related-posts` where they can be imported from JS and the file-loader while recognize and locate them.

I've added a simple README about the new `client/assets` directory.

Ready for some final reviews 🎉

simison reviewed (approved):
💯 

I really like the direction this is taking us with explicitly making sure assets really exist before using them.

I tested Calypso build, SDK dev/prod builds and all works great.

tyxla reviewed (approved):
Tests well 👍Code looks good too

I'm not happy to duplicate the images, but considering that we're using them in a different way, and that the related posts images might get soon updated, let's 🚢
